### PR TITLE
introduce protected_exprt

### PR DESCRIPTION
**Pull request by kroening**
This will serve as the interface that we aspire exprt to offer; in
particular, access to opX is protected.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cd7d05d184b7669e6c8f7b0da18f0f8c0f3118fa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97934741
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/pointer-analysis/value_set_fi.cpp
Comment (thk123) [inline src/pointer-analysis/value_set_fi.cpp]: :question: Why remove the type from lhs? Because we call `make_typecast` which makes this not an `index_exprt` :scream: -> :pick: don't change the type, instead create a typecast from the `index_exprt`

File: src/util/expr.h
Comment (thk123) [inline src/util/expr.h:327]: :question: Would it make more sense to use protected inheritance? I assume the answer is no as there remains calls to other methods, that have not been removed, but perhaps given this direction of travel it would making more sense to `using` them into the public section? (I think you can do this, but I have not actually tried).

Comment (thk123) [reply src/util/expr.h:327]: This would circumvent the need for this class entirely and just update the classes that inherit from `exprt` to protected extend it and in fact allow for explicitly identifying the classes that use the wrong interface.

File: src/util/expr.h
Comment (thk123) [inline src/util/expr.h]: :pick: I slightly worry about the name, in general, `X_exprt` means an expression representing an AST node of type X. I guess `protected` is never a node of a C/C++/Java AST, but perhaps this should be called `expr_protected` or `expr_api_protected` to indicate that this is an internal fudge rather than an actual GOTO thing.

**Review by thk123 — COMMENTED** [review]
This presumably requires a TG bump before merging - will create to see how much additional work is required.

**Comment by thk123** [comment]
Ah this doesn't yet compile on CBMC - will hold off on the bump front for now.

File: src/pointer-analysis/value_set_fi.cpp
Comment (kroening) [reply src/pointer-analysis/value_set_fi.cpp]: #3856 will fix this!

File: src/util/expr.h
Comment (kroening) [reply src/util/expr.h:327]: There are a lot of classes inheriting from `exprt`; furthermore, the interface of `exprt` should change as well, once we are there.

File: src/util/expr.h
Comment (kroening) [reply src/util/expr.h]: ok

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5d451cc8212d424f82acfbc6a736b5d0dc432106).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98429525
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by thk123** [comment]
@kroening This still isn't compiling on cbmc yet: 
```
value_set_fi.cpp:858:48: error: within this context
           index_expr.make_typecast(array.type());
```

So will hold off on the bump for now.

File: src/util/expr.h
Comment (tautschnig) [inline src/util/expr.h]: What is a "protected" class? Who is being protected? From what risks are they being shielded? When should anyone use it? I think this comment needs to be reworded.

File: src/util/expr.h
Comment (tautschnig) [inline src/util/expr.h]: It seems rather weird that newly introduced code would be deprecated? Can we just not have these?

File: src/util/expr.h
Comment (tautschnig) [inline src/util/expr.h]: Adding commented-out code without comments is just raising questions, but not providing answers.

File: src/util/pointer_offset_size.cpp
Comment (tautschnig) [inline src/util/pointer_offset_size.cpp]: I don't think that's an improvement?

File: src/util/expr.h
Comment (kroening) [reply src/util/expr.h]: sure, removed

File: src/util/expr.h
Comment (kroening) [reply src/util/expr.h]: sure, removed

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 7b6444ad7f1c7a74b2413cd8fbf39251ab54e8a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98484788
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d9ae0f752817912ddfc5b7f72a8208400306e31c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98523776
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/util/pointer_offset_size.cpp
Comment (tautschnig) [inline src/util/pointer_offset_size.cpp]: Is this still necessary, given `.find` is no longer protected?

File: src/util/std_expr.h
Comment (tautschnig) [inline src/util/std_expr.h]: Should this really be done? A `multi_ary_exprt` can, but need not have this many operands? For an unidentified `multi_ary_exprt` the caller should always use `.operands()` and just operate on that data structure.

File: src/util/expr.h
Comment (hannes-steffenhagen-diffblue) [inline src/util/expr.h]: Does that mean that `exprt` subclasses that so far _only_ have these as their API will get updated (e.g. classes like `typecast_exprt` that only have `op0` rather than a named accessor?).

Also, if we're already going to do an API do-over, why make the "new" expr inherit from `irept` at all, rather than have an irep as a protected field or something?

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
:thinking:

File: src/util/pointer_offset_size.cpp
Comment (kroening) [reply src/util/pointer_offset_size.cpp]: removed

File: src/util/std_expr.h
Comment (kroening) [reply src/util/std_expr.h]: replaced by ranged-checked opX().

File: src/util/expr.h
Comment (tautschnig) [reply src/util/expr.h]: I'm a bit confused, an intermittent version had more elaborate documentation. Do a `git reflog` to look at those?

File: src/util/expr.h
Comment (tautschnig) [reply src/util/expr.h]: Back again?

**Review by tautschnig — CHANGES_REQUESTED** [review]
It looks like some earlier version has been restored? Might be worth looking at the commits the GitHub lists or going for `git reflog`.

**Comment by kroening** [comment]
Back again!

File: src/util/std_expr.h
Comment (tautschnig) [inline src/util/std_expr.h:755]: Is this still needed?

**Review by tautschnig — APPROVED** [review]

File: src/util/std_expr.h
Comment (kroening) [reply src/util/std_expr.h:755]: Yes, there are still too many users -- but I'd say that it makes sense to work towards removing these two lines.

Comment (kroening) [reply src/util/std_expr.h:755]: (But at least it's not dangerous to use them).

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: a48f855762bb37d2a510ceddb1ee6d5939b45e2e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98702385
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/util/std_expr.h
Comment (tautschnig) [reply src/util/std_expr.h:755]: I don't understand - these are being deleted, even though `protected_exprt` makes them unavailable anyway.

Comment (kroening) [reply src/util/std_expr.h:755]: Ah, ok, the "this" was unclear.

The deletion of op2() and op3() is indeed no longer needed.

**Review by martin-cs — APPROVED** [review]
The changes to bits of code I am responsible for are fine.  I will leave it to others to have opinions on the exact API.  Being cleaner about the use of exprts / irepts is a good thing in general.

File: src/util/expr.h
Comment (peterschrammel) [inline src/util/expr.h:323]: Typo in `deprecated`

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
